adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +40,129 @@ def as_dict(self: Self) -> dict: ...
+            match status_code:
+                case 400:
+                    return BadRequestError
+                case 401:
+                    return UnauthorizedError
+                case 404:
+                    return NotFoundError
+                case 409:
+                    return ConflictError
+
+    return ResponseError
+
+
+def convert_exceptions(func: Callable) -> Callable:
+    async def wrapper(*arg: Iterable, **kwargs: dict) -> httpx.Response:
+        response = await func(*arg, **kwargs)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            raise _get_error_class(status_code=response.status_code) from e
+
+        return response
+
+    return wrapper
 
 
 class DefaultRequester(Requester):
-    def __init__(self: Self) -> None: ...
+    __slots__ = ("api_root", "session")
+
+    def __init__(self: Self, url: str, credentials: Credentials | None = None) -> None:
+        self.api_root = urljoin(url, "/api/v2/")
+        self.client = httpx.AsyncClient()
+        if credentials:
+            get_event_loop().run_until_complete(self.login_as(credentials=credentials))
+
+    async def login_as(self: Self, credentials: Credentials) -> None:
+        login_url = urljoin(self.api_root, "login/")
+        response = await self._do_request(self.client.post(url=login_url, data=credentials.dict()))
+
+        if response.status_code != 200:
+            message = f"Authentication error: {response.status_code} for url: {login_url}"
+            raise SessionError(message)
+
+    async def get(self: Self, path: str, query_params: dict | None = None) -> HTTPXRequesterResponse:
+        return await self.request(method="get", path=path, params=query_params or {})
+
+    async def post(self: Self, path: str, data: dict, **kwargs: dict) -> HTTPXRequesterResponse:
+        return await self.request(method="post", path=path, data=data, **kwargs)
+
+    async def patch(self: Self, path: str, data: dict, **kwargs: dict) -> HTTPXRequesterResponse:
+        return await self.request(method="patch", path=path, data=data, **kwargs)
+
+    async def delete(self: Self, path: str, **kwargs: dict) -> HTTPXRequesterResponse:
+        return await self.request(method="delete", path=path, **kwargs)
+
+    async def request(self: Self, method: str, path: str, **kwargs: dict) -> HTTPXRequesterResponse:
+        if not path.endswith("/"):
+            path = f"{path}/"
+        url = urljoin(self.api_root, path)
+        request_method = getattr(self.client, method.lower())
+
+        response = await self._do_request(request_method(url, headers=kwargs.pop("headers", {}), **kwargs))
+
+        return HTTPXRequesterResponse(response=response)
 
-    @property
-    async def session(self: Self) -> AsyncGenerator[Session]: ...
+    @convert_exceptions
+    async def _do_request(self: Self, request_coro: Coroutine[Any, Any, httpx.Response]) -> httpx.Response:
+        return await request_coro
```

Comment:
is this method required? looks like  it's the same as `request`

Reply:
it is used in `self.login()` and in `self.request()` which must handle exceptions differently